File: core/src/main/java/com/infoa2/core/widget/WidgetUtils.java
package com.infoa2.core.widget;


import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ScrollView;
import androidx.appcompat.app.AppCompatActivity;
import com.infoa2.core.widget.interfaces.ComponentActions;
import com.infoa2.core.widget.interfaces.FieldValidator;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by caio on 23/02/17.
 */

public class WidgetUtils {

    private static final String TAG = WidgetUtils.class.getName();
    protected static AppCompatActivity appCompatActivity;
    private static List<FieldValidator> fieldValidators;

    public static void init(AppCompatActivity appCompatActivity){
        WidgetUtils.appCompatActivity = appCompatActivity;
        fieldValidators = new ArrayList<>();
        findAllFieldValidators((ViewGroup) appCompatActivity.findViewById(android.R.id.content));

    }

    public static void init(ViewGroup viewGroup){
        fieldValidators = new ArrayList<>();
        findAllFieldValidators(viewGroup);

    }

    private static void findAllFieldValidators(ViewGroup viewGroup){
        int count = viewGroup.getChildCount();
        for (int i = 0; i < count; i++) {
            View view = viewGroup.getChildAt(i);
            if (view instanceof ViewGroup && !(view instanceof FieldValidator))
                findAllFieldValidators((ViewGroup) view);
            else if (view instanceof FieldValidator && view.getVisibility() != View.GONE) {
                fieldValidators.add((FieldValidator) view);

            }
        }

    }

    public static Boolean validateComponents() {
        Boolean isValid = Boolean.TRUE;
        for (FieldValidator fieldValidator : fieldValidators) {
            if(!fieldValidator.isValid()){
                isValid = Boolean.FALSE;
                Log.d("Widget Validator", fieldValidator.toString());
            }
        }
        return isValid;

    }

    public static Boolean validateComponents(ScrollView scrollView) {
        Object firstInvalidComponent = null;
        for (FieldValidator fieldValidator : fieldValidators) {
            if(!fieldValidator.isValid()){
                firstInvalidComponent = fieldValidator;
            }
        }
        if(firstInvalidComponent != null ){
            try {
                ((EditText) firstInvalidComponent).getParent().requestChildFocus((View) firstInvalidComponent, (View) firstInvalidComponent);
            }catch (Exception e){
                e.printStackTrace();
            }

        }
        return firstInvalidComponent  == null ? Boolean.TRUE : Boolean.FALSE;

    }

    public static void cleanComponents(List<ComponentActions> components) {
        for (ComponentActions componentActions : components) {
            componentActions.cleanComponent();
        }
    }

    public static void disableAndCleanComponents(List<ComponentActions> components, ComponentActions afterComponent) {
        int index = components.indexOf(afterComponent);
        List<ComponentActions> componentsFiltered = components.subList(index, components.size());
        for (ComponentActions componentActions : componentsFiltered) {
            componentActions.cleanComponent();
            componentActions.enableComponent(false);

        }
    }

    public static void disableComponents(List<ComponentActions> components, ComponentActions afterComponent) {
        int index = components.indexOf(afterComponent);
        List<ComponentActions> componentsFiltered = components.subList(index, components.size());
        for (ComponentActions componentActions : componentsFiltered) {
            componentActions.enableComponent(false);
        }
    }
}

File: core/src/main/java/com/infoa2/core/widget/WidgetUtils.kt
package com.infoa2.core.widget


import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.ScrollView
import androidx.appcompat.app.AppCompatActivity
import com.infoa2.core.widget.interfaces.ComponentActions
import com.infoa2.core.widget.interfaces.FieldValidator

import java.util.ArrayList

/**
 * Created by caio on 23/02/17.
 */

object WidgetUtils {

    private val TAG = WidgetUtils::class.java!!.getName()
    internal lateinit var appCompatActivity: AppCompatActivity
    private var fieldValidators: MutableList<FieldValidator>? = null

    fun init(appCompatActivity: AppCompatActivity) {
        WidgetUtils.appCompatActivity = appCompatActivity
        fieldValidators = ArrayList()
        findAllFieldValidators(appCompatActivity.findViewById<View>(android.R.id.content) as ViewGroup)

    }

    fun init(viewGroup: ViewGroup) {
        fieldValidators = ArrayList()
        findAllFieldValidators(viewGroup)

    }

    private fun findAllFieldValidators(viewGroup: ViewGroup) {
        val count = viewGroup.childCount
        for (i in 0 until count) {
            val view = viewGroup.getChildAt(i)
            if (view is ViewGroup && view !is FieldValidator)
                findAllFieldValidators(view)
            else if (view is FieldValidator && view.visibility != View.GONE) {
                fieldValidators!!.add(view as FieldValidator)

            }
        }

    }

    fun validateComponents(): Boolean? {
        var isValid: Boolean? = java.lang.Boolean.TRUE
        for (fieldValidator in fieldValidators!!) {
            fieldValidator.isValid().let {
                if (it) {
                    isValid = java.lang.Boolean.FALSE
                    Log.d("Widget Validator", fieldValidator.toString())
                }
            }
        }
        return isValid

    }

    fun validateComponents(scrollView: ScrollView): Boolean? {
        var firstInvalidComponent: Any? = null
        for (fieldValidator in fieldValidators!!) {
            fieldValidator.isValid().let {
                if (!it) {
                    firstInvalidComponent = fieldValidator
                }

            }
        }
        if (firstInvalidComponent != null) {
            try {
                (firstInvalidComponent as EditText).parent.requestChildFocus(
                    firstInvalidComponent as View?,
                    firstInvalidComponent as View?
                )
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
        return if (firstInvalidComponent == null) java.lang.Boolean.TRUE else java.lang.Boolean.FALSE

    }

    fun cleanComponents(components: List<ComponentActions>) {
        for (componentActions in components) {
            componentActions.cleanComponent()
        }
    }

    fun disableAndCleanComponents(components: List<ComponentActions>, afterComponent: ComponentActions) {
        val index = components.indexOf(afterComponent)
        val componentsFiltered = components.subList(index, components.size)
        for (componentActions in componentsFiltered) {
            componentActions.cleanComponent()
            componentActions.enableComponent(false)

        }
    }

    fun disableComponents(components: List<ComponentActions>, afterComponent: ComponentActions) {
        val index = components.indexOf(afterComponent)
        val componentsFiltered = components.subList(index, components.size)
        for (componentActions in componentsFiltered) {
            componentActions.enableComponent(false)
        }
    }
}
